### DISCUSSION TITLE
Random Sampling for imbalanced datasets

### DISCUSSION THREAD

**Discussion by ZahirBilal** [online-ml/river#1004] [Q&A] [answered]
Hello guys,

I am doing a research on performance of classifiers on extremely imbalanced datasets. For this use-case I tested almost all classifiers on River and also MOA (https://github.com/Waikato/moa).

So far the only algorithms that performed ok were the Random Sampling (both up and down) as well as the UOB algorithms in MOA.

My question is regarding the Random Sampling algorithms, was it developed based on some literature (paper?)? and if yes, please provide the publication.

Also, could someone please explain what is the parameter (`sampling_rate`) in `class RandomSampler(ClassificationSampler) `represent? 

Best wishes,
Zahir

**Comment by MaxHalford** [comment, marked as answer]
Hey there.

> I am doing a research on performance of classifiers on extremely imbalanced datasets. For this use-case I tested almost all classifiers on River and also MOA (https://github.com/Waikato/moa).

Some of the people behind MOA are the same ones behind scikit-multiflow, which merged with creme to become River. It's a small world :)

> My question is regarding the Random Sampling algorithms, was it developed based on some literature (paper?)? and if yes, please provide the publication.

No, not really. I got the idea from reading [Oza and Russell's online bagging and boosting paper](https://proceedings.mlr.press/r3/oza01a.html). I then realized that [rejection sampling](https://www.wikiwand.com/en/Rejection_sampling) was the mechanism I needed to sample data online with a desired target distribution. I wrote a [blog post](https://maxhalford.github.io/blog/undersampling-ratios/) about it, which led to the `RandomUnderSampler` implementation. Then it wasn't difficult to implement `RandomOverSampler`.

I can't find any of my notes concerning `RandomSampler`, but I vaguely remember doing the math by pen and paper. The `sampling_rate` simply indicates how much data to learn on. If `sampling_rate=0.5`, then the model wrapped by `RandomSampler` will be trained on 50% of the data.

I hope that helps.

**Reply by ZahirBilal** [reply]
Thanks Max for your help and infos. I got it completely now :)
